### PR TITLE
Feat(Core/Debugging): improve crash reports

### LINKED ISSUE #2351: [Core] No crash-reports being produced on crash
##### SMALL DESCRIPTION:
There are no .dmp and .txt files generated in the /Crashes/ folder anymore when the server crashes, this issue started sometime during the past 6 months.

##### STEPS TO REPRODUCE THE PROBLEM:
1. Open instance_molten_core.cpp
2. On line 57, inside the "OnPlayerEnter" function, add the code:
`Player* crash = NULL;`
`crash->SetAcceptWhispers(true);`
3. Build and run the server, in-game do ".gm on", ".tele moltencore", and then walk inside Molten Core, causing the server to crash.
4. Look in the /Crashes/ folder and see if any .dmp or .txt files were generated for the crash.

##### EXTRA NOTES:
So this issue started for me a couple of months ago when I rebased my project on latest AC at the time, though I didn't notice it until recently. I tried installing a complete fresh AC on latest revision and I experience the exact same issue there. I do have an older project thought that is based on an almost 1 year old version of AC, and that project works just fine in generating crash-reports. So something happened in AC recently that caused crash-reports to not be generated. 

It might be that this is config-related, and that somehow I need to enable this one my end. If that's the case I appreciate any help anyone can provide in pointing out how I enable it. I did ask about it on Discord and received no responses so I figured I'd report it here instead.

##### BRANCH(ES):
master

##### AC HASH/COMMIT:
003e8fa4598bcbbc28367b752c1183fed84a72f8

##### OPERATING SYSTEM:
Windows 10, latest updates.

##### MODULES:
None, able to reproduce on completely fresh AC.

##### OTHER CUSTOMIZATIONS:
None, able to reproduce on completely fresh AC.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/81962809-core-no-crash-reports-being-produced-on-crash?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
##### CHANGES PROPOSED:
- fix (enable) crash report in win
- Port all debugging scripts from [TC 3.3.5](https://github.com/TrinityCore/TrinityCore/tree/3.3.5/src/common/Debugging)

##### ISSUES ADDRESSED:
- Closes #2351 

##### TESTS PERFORMED:
- AzerothCore rev. 2d9d36718be5 2019-10-20 04:36:28 +0700 (debug branch) (Win64, RelWithDebInfo) at VS 16.3.5

##### HOW TO TEST THE CHANGES:
- Added test module https://github.com/Winfidonarleyan/mod-test-crash
- Write in-game `.test crash`
- Check core folder
For example:
image
- Open the folder `Crashes`
- Open last `.txt` file example: `2d9d36718be5_worldserver.exe_[20-10_18-15-10].txt`
- Check the first 51 lines
For example:

```
Revision: AzerothCore rev. 2d9d36718be5 2019-10-20 04:36:28 +0700 (debug branch) (Win64, RelWithDebInfo)
Date 20:10:2019. Time 18:15 
//=====================================================
*** Hardware ***
Processor: AMD Ryzen 5 3550H with Radeon Vega Mobile Gfx  
Number Of Processors: 8
Physical Memory: 6223524 KB (Available: 673436 KB)
Commit Charge Limit: 18282148 KB

*** Operation System ***
Windows 10 Home Edition (Version 10.0, Build 18362)

//=====================================================
Exception code: C0000005 ACCESS_VIOLATION
Fault address:  00007FF6652CE7C8 01:000000000063D7C8 D:\Core\Azeroth_Clean\worldserver.exe

Registers:
RAX:0000000000000000
RBX:0000000000000000
RCX:0000000000002002
RDX:00007FFBD281B770
RSI:000001969E4AF8B0
RDI:0000000000000000
R8: 00000196C83AC745
R9: 0000000000000000
R10:00000055EEBFEF40
R11:FFFFFFFFFFFFFFFF
R12:000001969E55D708
R13:00000055EEBFF588
R14:0000000000000000
R15:0000000000000001
CS:RIP:0033:00007FF6652CE7C8
SS:RSP:002B:00000000EEBFF0E0  RBP:EEBFF220
DS:002B  ES:002B  FS:0053  GS:002B
Flags:00010202

Call stack:
Address   Frame     Function      SourceFile
00007FF6652CE7C8  00000055EEBFF100  TestCrash_CS::HandleTestCrashCommand+18  D:\Git\azerothcore-wotlk\modules\mod-test-crash\src\TestCrash_SC.cpp line 34
00007FF664FED40E  00000055EEBFF2B0  ChatHandler::ExecuteCommandInTable+40E  D:\Git\azerothcore-wotlk\src\server\game\Chat\Chat.cpp line 300
00007FF664FED7C8  00000055EEBFF460  ChatHandler::ExecuteCommandInTable+7C8  D:\Git\azerothcore-wotlk\src\server\game\Chat\Chat.cpp line 277
00007FF664FEE690  00000055EEBFF4F0  ChatHandler::ParseCommands+F0  D:\Git\azerothcore-wotlk\src\server\game\Chat\Chat.cpp line 430
00007FF66523DD92  00000055EEBFF670  WorldSession::HandleMessagechatOpcode+712  D:\Git\azerothcore-wotlk\src\server\game\Handlers\ChatHandler.cpp line 297
00007FF664DD5229  00000055EEBFF7A0  WorldSession::Update+499  D:\Git\azerothcore-wotlk\src\server\game\Server\WorldSession.cpp line 323
00007FF664D44BC6  00000055EEBFF850  World::UpdateSessions+326  D:\Git\azerothcore-wotlk\src\server\game\World\World.cpp line 2636
00007FF664D43AB1  00000055EEBFF8E0  World::Update+2D1  D:\Git\azerothcore-wotlk\src\server\game\World\World.cpp line 2170
00007FF664D258B0  00000055EEBFF9A0  WorldRunnable::run+120  D:\Git\azerothcore-wotlk\src\server\worldserver\WorldThread\WorldRunnable.cpp line 49
00007FF6652942CF  00000055EEBFF9D0  std::thread::_Invoke<std::tuple<void (__cdecl*)(void *),void *>,0,1>+F  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\thread line 40
00007FFBD2750E72  00000055EEBFFA00  _beginthreadex+142
00007FFBD3987BD4  00000055EEBFFA30  BaseThreadInitThunk+14
00007FFBD4B0CED1  00000055EEBFFAB0  RtlUserThreadStart+21
```

##### KNOWN ISSUES AND TODO LIST:

- [ ] Test in Linux
- [ ] Test in Windows
- [ ] Test in Docker

##### Target branch(es):
- [x] Master

## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by rbedfordpro** [comment]
Will this work for when one is using the docket-compose?

**Comment by FrancescoBorzi** [comment]
@rbedfordpro I don't see why not. When you use Docker, you basically run a normal AC instance inside a linux container. So everything should just work fine.

**Comment by Winfidonarleyan** [comment]
Added detailed info

**Comment by Malow** [comment]
Tested forced hard-coded crashes in the following scenarios in Debug-mode:
- 	When you invite someone to a group (Using a PlayerScript)
- 	When you use an item (inside CastItemUseSpell)
- 	When you type .server info (from cs_server)
- 	When you enter MoltenCore (from an instance-script)
- 	When Onyxia resets (from a ScriptedAI)

All of them perfectly generates crash reports.

I did however notice that when I try it out in Release-mode I do get a crash-report, but it contains mostly nonsense. I don't know if that is by design due to limitations when building in Release mode though. As an example here's the exact same type of crash in Debug and in Release:
Debug: https://pastebin.com/nthFM9fZ
Release: https://pastebin.com/bCapUfEe
As you can see the Release one is pretty much useless since it doesn't contain any info about "CastItemUseSpell", which is what is causing the crash here.

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan can you fix the Codacy errors?

@Malow thanks for testing, I think that is the intended behaviour

**Comment by Winfidonarleyan** [comment]
@FrancescoBorzi need help for Codacy

File: src/common/Debugging/Errors.cpp [hunk 1]
```diff
@@ -1,60 +1,105 @@
...
+        vsnprintf(&formatted[0], length + 1, format, args);
+
+        return formatted;
+    }
+}
+
+namespace Trinity
+{
+
+void Assert(char const* file, int line, char const* function, std::string debugInfo, char const* message)
```
Comment (Viste) [inline src/common/Debugging/Errors.cpp]: ```void Assert(char const* file, int line, char const* function, std::string const& debugInfo, char const* message)```

File: src/common/Debugging/WheatyExceptionReport.cpp [hunk 2]
```diff
@@ -1170,26 +1385,68 @@ WheatyExceptionReport::GetBasicType(DWORD typeIndex, DWORD64 modBase)
...
+        va_end(argptr);
+    }
+    else
+    {
+        retValue = StackLog(format, argptr);
+        va_end(argptr);
+    }
+
+    return retValue;
+}
```
Comment (Viste) [inline src/common/Debugging/WheatyExceptionReport.cpp]: ```va_end(argptr);```
must be after if else and before 
```return retValue;```

**Comment by Viste** [comment]
also need this: https://github.com/Viste/trinity/commit/37cba47283a1d8f1eda12ca07a5d6ca37832240f#diff-f547c8642582953a953957d6baefea4b

**Comment by Winfidonarleyan** [comment]
Fix some codacy warnings

**Comment by Viste** [comment]
> Fix some codacy warnings

all others warnings not from this PR :)

**Comment by FrancescoBorzi** [comment]
is this still WIP or ready to be tested?

**Comment by vhiperdev** [comment]
Tested Windows reports received on [Debug and Release] Ok

I have not tested RelWithDebInfo

**Comment by Winfidonarleyan** [comment]
Tranks for test @vhiperdev

**Pull request merged**